apps/eth_blockchain/lib/eth_blockchain/transaction.ex

```diff
@@ -0,0 +1,261 @@
+  Possible combinaisons of tuple elements:
+  /3 {from_addr, to_addr, amount}
+    -> will transfer `amount` ether from `from_addr` to `to_addr` with the default gas price
+  /4 {from_addr, to_addr, amount, gas_price}
+    -> will transfer `amount` ether from `from_addr` to `to_addr` with the specified gas price
+  /4 {from_addr, to_addr, amount, contract_addr}
+    -> will transfer `amount` token using the ERC20 contract residing at `contract_addr`
+    from `from_addr` to `to_addr` with the default gas price
+  /5 {from_addr, to_addr, amount, contract_addr, gas_price}
+    ->
+      if contract_addr is `0x0000000000000000000000000000000000000000`:
+        will transfer `amount` ether from `from_addr` to `to_addr` with the specified gas price
+      otherwise:
+        will transfer `amount` token using the ERC20 contract residing at `contract_addr`
+        from `from_addr` to `to_addr` with the specified gas price
+
+  Returns:
+  `{:ok, tx_hash}` if successful
+  or
+  `{:error, error_code}` or `{:error, error_code, message}` if failed
+  """
+  @spec send(tuple(), atom() | nil, pid() | nil) ::
+          {atom(), String.t()} | {atom(), atom()} | {atom(), atom(), String.t()}
+  def send(params, adapter \\ nil, pid \\ nil)
+
+  # send default (eth) without gas price
+  def send({from_address, to_address, amount}, adapter, pid) do
+    gas_price = Application.get_env(:eth_blockchain, :default_gas_price)
+
+    send(
+      {from_address, to_address, amount, "0x0000000000000000000000000000000000000000", gas_price},
+      adapter,
+      pid
+    )
+  end
+
+  # send default (eth) with gas price
+  def send({from_address, to_address, amount, gas_price}, adapter, pid)
+      when is_integer(gas_price) do
+    send(
+      {from_address, to_address, amount, "0x0000000000000000000000000000000000000000", gas_price},
+      adapter,
+      pid
+    )
+  end
+
+  # send eth with gas price
+  def send(
+        {from_address, to_address, amount, "0x0000000000000000000000000000000000000000",
+         gas_price},
+        adapter,
+        pid
+      ) do
+    gas_limit = Application.get_env(:eth_blockchain, :default_eth_transaction_gas_limit)
+
+    case get_transaction_count({from_address}, adapter, pid) do
+      {:ok, nonce} ->
+        %__MODULE__{
+          gas_limit: gas_limit,
+          gas_price: gas_price,
+          nonce: int_from_hex(nonce),
+          to: from_hex(to_address),
+          value: amount
+        }
+        |> sign_and_hash(from_address)
+        |> send_raw(adapter, pid)
+
+      error ->
+        error
+    end
+  end
+
+  # send to contract address without gas price
+  def send({from_address, to_address, amount, contract_address}, adapter, pid)
+      when is_binary(contract_address) do
+    gas_price = Application.get_env(:eth_blockchain, :default_gas_price)
+    send({from_address, to_address, amount, contract_address, gas_price}, adapter, pid)
```

Comment:
Might be able to simplify default gas price retrieving with (for both eth/contract):

```
def send({_, _, _, _} = t, adapter, pid) do
  gas_price = Application.get_env(:eth_blockchain, :default_gas_price)
  t2 = Tuple.append(t, gas_price)
  send(t2, adapter, pid)
end
```

Reply:
Alternatively, it might be easier if we use map instead of tuple.